aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -685,6 +685,513 @@ Tensor& matmul_out(Tensor &result, const Tensor & tensor1, const Tensor & tensor
+// compute the matrix exponentials of different orders, i.e.
+// first `num_matrices` matrices from the list l := {I, A, A^2, A^3, A^6}
+// in a contiguous block of memory such that
+// buffer[0, ...] = l[0], // I
+// buffer[1, ...] = l[1], // A
+// ...
+// buffer[num_matrices - 1, ...] = l[num_matries - 1]
+void _fill_matrix_powers(Tensor& buffer, const Tensor& a, int num_matrices) {
+  // fill I
+  buffer.select(0, 0).copy_(
+    at::eye(a.size(-1), buffer.options()).expand_as(a)
+  );
+
+  // fill a
+  buffer.select(0, 1).copy_(a);
+
+  // fill a^2
+  if (2 <= num_matrices - 1) {
+    at::native::matmul(
+      buffer.select(0, 2), // out for a^2
+      buffer.select(0, 1),
+      buffer.select(0, 1)
+    );
+  }
+
+  // fill a^3
+  if (3 <= num_matrices - 1) {
+    at::native::matmul(
+      buffer.select(0, 3), // out for a^3
+      buffer.select(0, 1),
+      buffer.select(0, 2)
+    );
+  }
+
+  // fill a^6
+  if (4 <= num_matrices - 1) {
+    at::native::matmul(
+      buffer.select(0, 4),
+      buffer.select(0, 3),
+      buffer.select(0, 3)
+    );
+  }
+}
+
+inline Tensor _pin_and_move_memory_if_cuda_input(
+  const Tensor& mem,
+  const Tensor& in
+) {
+  return (in.device().type() == at::kCUDA)
+    ? mem.pin_memory().to(at::device_of(in).value(), /*non_blocking=*/true)
+    : mem;
+}
+
+// convert a 1D blob to a 2D Tensor of size [1, blob.size()]
+// such that blob.device() == in.device())
+// designed to be used with _compute_linear_combination
+template <typename scalar_t>
+inline Tensor _blob_to_Tensor(
+  std::initializer_list<scalar_t> blob,
+  const Tensor& in
+) {
+  // we convert to void* expecitly because begin() returns
+  // a pointer to a constant.
+  // Blob is assumed to be a 1D array, that is why
+  // we also insert a fake dimension so that the result could directly
+  // be used in _compute_linear_combination
+  auto tensor = at::from_blob((void*)blob.begin(), blob.size(), in.dtype())
+    .unsqueeze(0);
+  return _pin_and_move_memory_if_cuda_input(tensor, in);
+}
+
+// I + A
+Tensor compute_T1(const Tensor& A) {
+  // 2 for {I, A}
+  auto As = _allocate_buffer(A, 2);
+  _fill_matrix_powers(As, A, 2);
+  return As.sum(0);
+}
+
+// I + A + A^2 / 2
+Tensor compute_T2(const Tensor& A) {
+  auto As = _allocate_buffer(A, 3);
+  // 3 for {I, A, A^2}
+  _fill_matrix_powers(As, A, 3);
+  As.select(0, 2).div_(2.0);
+  return As.sum(0);
+}
+
+// I + A + A^2 * (I / 2 + A / 6 + A^2 / 24)
+template <typename scalar_t>
+Tensor compute_T4(const Tensor& A) {
+  auto As = _allocate_buffer(A, 4);
+  // 3 for {I, A, A^2}
+  _fill_matrix_powers(As, A, 3);
+  
+  at::native::matmul(
+    // output for A^2 * (I / 2 + A / 6 + A^2 / 24)
+    As.select(0, 3),
+    // contains A^2
+    As.select(0, 2),
+    // computes (I / 2 + A / 6 + A^2 / 24)
+    at::native::_compute_linear_combination(
+      As.narrow(0, 0, 3),
+      _blob_to_Tensor<scalar_t>({1 / 2.0, 1 / 6.0, 1 / 24.0}, A)
```

Comment:
Fun fact: when you launch CUDA kernels, you get some space to transmit arbitrary data from CPU. This means that if you write a kernel that, e.g., multiplies a tensor by a scalar, it is better to send the scalar via this mechanism, rather than making it a CUDA tensor and doing a host-to-device copy. This seems a bit applicable here.

However, this fact isn't so useful here because I can't think of any way you can conveniently take advantage of this without having to write a bunch of manual kernels, and in an earlier comment I told you to implement `_compute_linear_combination` as matmul.

Reply:
Another problem is that launching kernels one by one will require synchronization and atomics for CUDA, which result in non-determinism. Now, backward of the matrix exponential is again a matrix exponential of some input. In my tests autograd did not tolerate non-determinism in backward at all.

Reply:
Hmm, maybe it can be done, if kernel supports `out` parameter. The kernels are launched in the order of their invocation, so the result should be deterministic. Determinism is important because the backward of a matrix exponential is a matrix exponential in this PR. And in my tests autograd did not tolerate non-determinism in the backward tests at all.

Reply:
out parameter is a different thing; that just says where the results should get written to. I'm actually talking about the protocol by which CUDA launches kernels

Reply:
Well, I suppose 'mul' does support this m mechanism, no? Then a linear combination x_1 A_1 + etc can be written as a superposition of 'mul_out' and 'sum_out', that is what I meant.